src/py21cmfast/drivers/coeval.py

```diff
@@ -0,0 +1,901 @@
+
+        return cls.from_file(fname, load_data=load_data)
+
+    def gather(
+        self,
+        fname: str | None | Path = None,
+        kinds: Sequence | None = None,
+        clean: bool | dict = False,
+        direc: str | Path | None = None,
+    ) -> Path:
+        """Gather the cached data associated with this object into its file."""
+        kinds = kinds or [
+            "init",
+            "perturb_field",
+            "halobox",
+            "ionized_box",
+            "spin_temp",
+            "brightness_temp",
+        ]
+
+        clean = kinds if clean and not hasattr(clean, "__len__") else clean or []
+        if any(c not in kinds for c in clean):
+            raise ValueError(
+                "You are trying to clean cached items that you will not be gathering."
+            )
+
+        direc = Path(direc or config["direc"]).expanduser().absolute()
+        fname = Path(fname or self.get_unique_filename()).expanduser()
+
+        if not fname.exists():
+            fname = direc / fname
+
+        for kind in kinds:
+            redshifts = (f[0] for f in self.cache_files[kind])
+            for i, z in enumerate(redshifts):
+                cache_fname = self.cache_files[kind][i][1]
+
+                obj = self.get_cached_data(kind, redshift=z, load_data=True)
+                with h5py.File(fname, "a") as fl:
+                    cache = (
+                        fl.create_group("cache") if "cache" not in fl else fl["cache"]
+                    )
+                    kind_group = (
+                        cache.create_group(kind) if kind not in cache else cache[kind]
+                    )
+
+                    zstr = f"z{z:.2f}"
+                    if zstr not in kind_group:
+                        z_group = kind_group.create_group(zstr)
+                    else:
+                        z_group = kind_group[zstr]
+
+                    obj.write_data_to_hdf5_group(z_group)
+
+                    if kind in clean:
+                        os.remove(cache_fname)
+        return fname
+
+    def _get_prefix(self):
+        return "{name}_z{redshift:.4}_{{hash}}_r{seed}.h5".format(
+            name=self.__class__.__name__,
+            redshift=float(self.redshift),
+            seed=self.random_seed,
+        )
+
+    def _input_rep(self):
+        return "".join(
+            repr(getattr(self, inp))
+            for inp in [
+                "user_params",
+                "cosmo_params",
+                "astro_params",
+                "flag_options",
+                "global_params",
+            ]
+        )
+
+    def get_unique_filename(self):
+        """Generate a unique hash filename for this instance."""
+        return self._get_prefix().format(
+            hash=md5((self._input_rep() + self._particular_rep()).encode()).hexdigest()
+        )
+
+    def _write(self, direc=None, fname=None, clobber=False):
+        """
+        Write the lightcone to file in standard HDF5 format.
```

Comment:
minor note to examine some of the older docstrings for mistakes, in this case lightcone/coeval mixup